src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp

```diff
@@ -81,6 +101,13 @@ IMergingAlgorithm::Status ReplacingSortedAlgorithm::merge()
         {
             max_pos = current_pos;
             setRowRef(selected_row, current);
+            if (cleanup)
+                if (assert_cast<const ColumnUInt8 &>(*(*selected_row.all_columns)[is_deleted_column_number]).getData()[selected_row.row_num])
+                    selected_row.clear();
+        }
+        else if (cleanup && is_deleted)
+        {
+                current_row.clear();
```

Comment:
Is `cleanup` set to true when doing SELECT with FINAL?

Reply:
cleanup is by default set to false; when doing SELECT ... FINAL (and if the settings is set to 'never') the cleanup wont be applied (as shown in the tests). the CLEANUP keyword has to be added to be triggered (or set the setting to 'Always').

Reply:
> when doing SELECT ... FINAL (and if the settings is set to 'never') the cleanup wont be applied (as shown in the tests)

I think SELECT ... FINAL should not show deleted rows

Reply:
I implement that change in my last commit (with tests) [13cf434](https://github.com/ClickHouse/ClickHouse/pull/41005/commits/13cf434a0ced92cd5f0e7d55f31b2a712b4f56df)
I just want to be sure that we are all on the same page (@tavplubix @filimonov @gontarzpawel). 
Because now `SELECT ... FINAL` doesn't return rows with `is_deleted = true`. So, the only way to have all rows (even those flagged as is_deleted) without duplicates with a `SELECT` is to run `OPTIMIZE` first. 
I just want to be sure that we all agree on the behaviour.

Reply:
I meant that `SELECT ... FINAL` should choose the row with the greatest versions for each key at first and after that it should remove rows with `is_deleted`

Reply:
Ok so if I haven't missed something, the [commit f69ad50](https://github.com/ClickHouse/ClickHouse/commit/f69ad505c3974577cf98d025f92925d1bf39333b), should do that.